sql/expression/function/json_set.go

```diff
@@ -188,31 +182,104 @@ func (j *JSONSet) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
 // processPath checks the given json path for the correct mysql syntax, checks nested paths for their existence
 // in the provided json doc, determines whether JSON_SET will do nothing with no error for this path, and processes
 // the given json path to use the appropriate sjson syntax.
-func processPath(ctx *sql.Context, doc types.JSONDocument, path string, pass bool) (string, bool, error) {
-	// make sure path starts with '$'
+func processPath(doc, path string) (string, bool, error) {
+	// check path starts with '$'
 	if path[0] != '$' {
 		return "", false, fmt.Errorf("Invalid JSON path expression")
 	}
-
-	// no wildcards allowed in path
-	if strings.Contains(path, "*") || strings.Contains(path, "**") {
+	// check no wildcards in path
+	if strings.Contains(path, "*") {
 		return "", false, fmt.Errorf("Path expressions may not contain the * and ** tokens")
 	}
+	path = path[1:]
 
-	// if the nested path does not already exist, we do nothing
-	if len(strings.Split(path, ".")) > 2 {
-		extractedVal, _ := doc.Extract(ctx, path)
-		if extractedVal == nil {
-			pass = true
-		}
+	// tokenize each field of the path
+	var parsed parsedPath
+	parsed.parts = strings.Split(path, ".")
+	if parsed.parts[0] == "" {
+		parsed.parts = parsed.parts[1:]
 	}
 
-	// match mysql indexing syntax to sjson syntax
-	path = strings.ReplaceAll(path, "[", ".")
-	path = strings.ReplaceAll(path, "]", "")
+	// process each field of the path
+	for i, part := range parsed.parts {
+		formattedPart := part
+
+		// handle any indexing in this field
+		if strings.Contains(part, "[") {
+			// tokenize indexes
+			formattedPart = strings.ReplaceAll(formattedPart, "]", "")
+			indexedParts := strings.Split(formattedPart, "[")
+			if indexedParts[0] == "" {
+				indexedParts = indexedParts[1:]
+			}
+			formattedPart = ""
+
+			// process each token
+			for j, idxPart := range indexedParts {
+				if idx, err := strconv.Atoi(idxPart); err == nil { // if token is an int, it's an index
+					if j == 0 { // if this is the first token of the field
+						if i == 0 && path[1] == '.' { // path cannot start with index after '.'
+							return "", false, fmt.Errorf("Invalid JSON path expression")
+						} else if i == 0 && path[1] != '.' { // ordinal indexing
+							return "", false, fmt.Errorf("Ordinal indexing not supported")
+						}
```

Comment:
Just brainstorming... would a regex be able to catch this case? I'm wondering if this would be a little bit of logic/error-checking that you could pull up to the start of the function and out of this more complex processing loop to help simplify a little bit. (Just an idea; I haven't thought through whether it would work well or not.)